numpy/core/arrayprint.py

```diff
@@ -503,18 +518,25 @@ def array2string(a, max_line_width=None, precision=None,
     '[0x0L 0x1L 0x2L]'
 
     """
-    # Deprecation 05-16-2017  v1.14
-    if style is not np._NoValue:
-        warnings.warn("'style' argument is deprecated and no longer functional",
-                      DeprecationWarning, stacklevel=3)
+    legacy = kwarg.pop('legacy', None)
+    if kwarg:
+        msg = "array2string() got unexpected keyword argument '{}'"
+        raise TypeError(msg.format(kwarg.popitem()[0]))
 
     overrides = _make_options_dict(precision, threshold, edgeitems,
                                    max_line_width, suppress_small, None, None,
-                                   sign, formatter)
+                                   sign, formatter, legacy)
     options = _format_options.copy()
     options.update(overrides)
 
-    if a.size == 0:
+    if options['legacy'] and a.shape == () and not a.dtype.names:
+        if style is None:
+            style = repr
+        return style(a.item())
+
+    if style is not None and a.shape == ():
+        return style(a[()])
```

Comment:
I don't think this is recursion-safe for `class subclass(np.ndarray): pass; a = subclass()` - `[()]` doesn't return a scalar for subclasses.

Reply:
it seems to return a scalar here...
```python
[1]: class sub(np.ndarray):
...:     pass
...: 
[2]: a = np.array(1.0).view(sub)
[3]: a[()]
1.0
[4]: type(a[()])
numpy.float64
```

Reply:
Indeed, that is the default. I'm not sure how this would work with class which like `Quantity` explicitly circumvent this by overriding `__getitem__` (since it obviously would be horrible if just looking at an individual item would lose the unit... I think I may have mentioned I disliked the scalar classes? ;-).  

Now, since `Quantity` overrides `__str__` too, this is not an issue for it, but it adds to @eric-wieser's point that perhaps one should not count on `[()]` returning a different class. That said, in this situation the user is providing their own `style`, so at some level it is up to them to ensure it is a function that doesn't just call back.

Reply:
Well if we merge #9201 this problem disappears!